File: python/107-Binary Tree Level Order Traversal II.py
from typing import List
from bintree_helper import *

class Solution:
    # def levelOrderBottom(self, root) -> List[List[int]]:
    #     "DFS, Time: O(n), Space: O(n)"
    #     if root is None: return []
    #     result = []
    #     def dfs(node, level):
    #         if level == len(result):
    #             result.append([])
    #         result[level].append(node.val)
    #         if node.left: dfs(node.left, level+1)
    #         if node.right: dfs(node.right, level+1)
    #     dfs(root, 0)
    #     return result[::-1]
    def levelOrderBottom(self, root) -> List[List[int]]:
        "BFS, Time: O(n), Space: O(n)"
        from collections import deque
        if root is None: return []

        q = deque([root])
        result = []
        while q:
            n = len(q)
            result.append([])
            for _ in range(n):
                node = q.popleft()
                result[-1].append(node.val)
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)
        return result[::-1]


root = deserialize("[3,9,20,null,null,15,7]")
result = Solution().levelOrderBottom(root)
assert result == [
  [15,7],
  [9,20],
  [3]
]

File: cpp/107. Binary Tree Level Order Traversal II.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// BFS, Time: O(n), Space: O(n)
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        if(!root) return vvi();
        queue<TreeNode*> q; q.push(root);
        vvi ans;
        while(!q.empty()) {
            int sz = q.size();
            ans.push_back(vi(sz));
            for(int i = 0; i < sz; i++) {
                TreeNode *node = q.front(); q.pop();
                ans.back()[i] = node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    TreeNode *root = new TreeNode();
    Solution solution;
    vvi ans = solution.levelOrderBottom(root);
    for(auto l : ans) {
        cout << "[ ";
        for(auto num : l) {
            cout << num << " ";
        }
        cout << " ]";
    }
}